ukvm/ukvm_hv_kvm_aarch64.c

```diff
@@ -0,0 +1,137 @@
+ */
+
+/*
+ * ukvm_hv_kvm_x86_64.c: x86_64 architecture-dependent part of KVM backend
+ * implementation.
+ */
+
+#define _GNU_SOURCE
+#include <assert.h>
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "ukvm.h"
+#include "ukvm_hv_kvm.h"
+#include "ukvm_cpu_aarch64.h"
+
+void ukvm_hv_vcpu_init(struct ukvm_hv *hv, ukvm_gpa_t gpa_ep,
+        ukvm_gpa_t gpa_kend, char **cmdline)
+{
+    struct ukvm_hvb *hvb = hv->b;
+
+    /* Setup aarch64 guest memroy */
+    ukvm_aarch64_setup_memory(hvb->vmfd, hv->mem, 0, hv->mem_size);
+
+    /* Setup aarch64 system registers */
+    ukvm_aarch64_setup_system(hvb->vmfd, hvb->vcpufd);
+
+    /* Setup aarch64 guest vector table */
+    ukvm_aarch64_setup_vectors(hvb->vcpufd, gpa_ep - VBAR_TABLE_LEN);
+
+    /* Setup aarch64 core registers */
+    ukvm_aarch64_setup_core(hv, gpa_ep, gpa_kend);
+
+    /* Free the memeory regions list, we won't use it any more */
+    free(ukvm_region_list);
+
+    *cmdline = (char *)(hv->mem + AARCH64_CMDLINE_BASE);
+}
+
+void ukvm_hv_vcpu_loop(struct ukvm_hv *hv)
+{
+    struct ukvm_hvb *hvb = hv->b;
+    int ret;
+
+    while (1) {
+        ret = ioctl(hvb->vcpufd, KVM_RUN, NULL);
+        if (ret == -1 && errno == EINTR)
+            continue;
+        if (ret == -1) {
+            if (errno == EFAULT) {
+                uint64_t pc;
+                ret = ukvm_aarch64_dump_pc(hvb->vcpufd, &pc);
+                if (ret == -1)
+                    err(1, "KVM: Dump PC failed after guest fault");
+                errx(1, "KVM: host/guest translation fault: pc=0x%lx", pc);
+            }
+            else
+                err(1, "KVM: ioctl (RUN) failed");
+        }
+
+        int handled = 0;
+        for (ukvm_vmexit_fn_t *fn = ukvm_core_vmexits; *fn && !handled; fn++)
+            handled = ((*fn)(hv) == 0);
+        if (handled)
+            continue;
+
+        struct kvm_run *run = hvb->vcpurun;
+
+        switch (run->exit_reason) {
+        case KVM_EXIT_MMIO: {
+            if (!run->mmio.is_write || run->mmio.len != 4)
+                errx(1, "Invalid guest port access: port=0x%llx", run->mmio.phys_addr);
+
+            if (run->mmio.phys_addr < UKVM_HYPERCALL_PIO_BASE ||
+                    run->mmio.phys_addr >=
+                    (UKVM_HYPERCALL_PIO_BASE + (UKVM_HYPERCALL_MAX << 2)))
+                errx(1, "Invalid guest port access: port=0x%llx", run->mmio.phys_addr);
+
+            int nr = (run->mmio.phys_addr - UKVM_HYPERCALL_PIO_BASE) >> 2;
+
+            /* Guest has halted the CPU, this is considered as a normal exit. */
+            if (nr == UKVM_HYPERCALL_HALT)
+                return;
+
+            ukvm_hypercall_fn_t fn = ukvm_core_hypercalls[nr];
+            if (fn == NULL)
+                errx(1, "Invalid guest hypercall: num=%d", nr);
+
+            ukvm_gpa_t gpa = 0;
+            memcpy(&gpa, run->mmio.data, run->mmio.len);
```

Comment:
On ARM64, when I use a direct pointer run->mmio.data[8], I always got the gcc "strict aliasing".
In this case, I have to memcpy instead. But your comment is right, I should not use the run->mmio.len directly. I will fix it.

But gpa is a 64-bit data, so the upper 32-bit is random. This is
will cause the UKVM_CHECKED_GPA_P failed in some applications.
In order to address this issue, we reset the gpa to 0 before get
address from run->mmio.data. In this case, the unused 32-bit value
should be set to 0.